### PR TITLE
submit solution only if  solution is better than the solution in the chain

### PR DESCRIPTION
**Pull request by josojo**
As title

----
Testplan:
unit tests

### CODE REVIEW

File: driver/src/driver/stablex_driver.rs [hunk 1]
```diff
@@ -39,14 +39,21 @@ impl<'a> StableXDriver<'a> {
             solution
         };
 
-        let submitted = if solution.executed_sell_amounts.iter().sum::<u128>() > 0 {
-            self.contract.submit_solution(batch, orders, solution)?;
-            info!("Successfully applied solution to batch {}", batch);
-            true
+        let current_objective_value = self.contract.get_current_objective_value()?;
+        let submitted = if let Some(surplus) = solution.surplus {
```
Comment (josojo) [inline driver/src/driver/stablex_driver.rs]: I was searching here for a more elegant solution, but I can not find it.

Thread discussing it:
https://github.com/rust-lang/rfcs/issues/929#issuecomment-81355948

File: driver/src/driver/stablex_driver.rs [hunk 2]
```diff
@@ -39,14 +39,21 @@ impl<'a> StableXDriver<'a> {
...
+                info!("Successfully applied solution to batch {}", batch);
+                true
+            } else {
+                info!("Not submitting trivial solution for batch {}", batch);
+                false
+            }
         } else {
             info!("Not submitting trivial solution for batch {}", batch);
             false
         };
```
Comment (bh2smith) [inline driver/src/driver/stablex_driver.rs]: This looks like duplicate code. Was this intended?

File: driver/src/driver/stablex_driver.rs (hunk 1 above)
Comment (fleupold) [reply driver/src/driver/stablex_driver.rs]: You are comparing surplus (should probably be renamed to utility in a separate PR) with the current_objective_value (which at the moment is volume). I don't think this works in the current setting as a solution with little surplus would not be corrected as long as it had more volume than a new solution had utility. Can't we compute and compare the proper volume instead?

As for how to make this more elegant with less nested if statements, you could have a helper method that returns a `SolutionSubmission` enum (e.g. Trivial, NotGoodEnough, Submit) and then have a match statement with only one level.

Comment (josojo) [reply driver/src/driver/stablex_driver.rs]: I thought, although this is not the correct solution, it is better than checking just that is is not the trivial solution.
Not sure, whether it is worth to code the volume formula ( actually the fee formula), as the rounding that would be needed to be considered is non-trivial.

But yeah, I am happy to pause this PR and keep it for later, once Tom's solver and the smart contract calculate the same objective value.

File: driver/src/driver/stablex_driver.rs (hunk 2 above)
Comment (josojo) [reply driver/src/driver/stablex_driver.rs]: Yes, this is what I was referring to here:
```
I was searching here for a more elegant solution, but I can not find it.

Thread discussing it:
rust-lang/rfcs#929 (comment)
```

File: driver/src/driver/stablex_driver.rs [hunk 3]
```diff
@@ -41,14 +41,21 @@ impl<'a> StableXDriver<'a> {
...
-            info!("Successfully applied solution to batch {}", batch);
-            true
+        let current_objective_value = self.contract.get_current_objective_value()?;
+        let submitted = if let Some(objective_value) = solution.objective_value {
+            if current_objective_value < objective_value {
+                self.contract.submit_solution(batch, orders, solution)?;
+                info!("Successfully applied solution to batch {}", batch);
+                true
+            } else {
+                info!("Not submitting trivial solution for batch {}", batch);
```
Comment (fleupold) [inline driver/src/driver/stablex_driver.rs]: This log message seems off.

File: driver/src/driver/stablex_driver.rs [hunk 4]
```diff
@@ -41,14 +41,21 @@ impl<'a> StableXDriver<'a> {
             solution
         };
 
-        let submitted = if solution.executed_sell_amounts.iter().sum::<u128>() > 0 {
-            self.contract.submit_solution(batch, orders, solution)?;
-            info!("Successfully applied solution to batch {}", batch);
-            true
+        let current_objective_value = self.contract.get_current_objective_value()?;
+        let submitted = if let Some(objective_value) = solution.objective_value {
+            if current_objective_value < objective_value {
```
Comment (fleupold) [inline driver/src/driver/stablex_driver.rs]: How about an objective value of 0? Would that ever get submitted?

Comment (bh2smith) [reply driver/src/driver/stablex_driver.rs]: At the moment, the contract does not allow for trivial solution submission. Furthermore, it does not accept solutions with zero objective evaluation either.

File: driver/src/driver/stablex_driver.rs (hunk 3 above)
Comment (bh2smith) [reply driver/src/driver/stablex_driver.rs]: I agree. Also, this wouldn't be possible anyway.

File: driver/src/driver/stablex_driver.rs (hunk 1 above)
Comment (nlordell) [reply driver/src/driver/stablex_driver.rs]: You could use match statements with conditional patterns, something like this:
```rust
        let submitted = match solution.surplus {
            Some(surplus) if current_objective_value < surplus => {
                self.contract.submit_solution(batch, orders, solution)?;
                info!("Successfully applied solution to batch {}", batch);
                true
            },
            _ => {
                info!("Not submitting trivial solution for batch {}", batch);
                false
            },
        };
```

Comment (nlordell) [reply driver/src/driver/stablex_driver.rs]: You could also use `Option::map` and do something like:
```rust
        let submitted = if Some(true) = solution.surplus.map(|surplus| current_objective_value < surplus) {
            self.contract.submit_solution(batch, orders, solution)?;
            info!("Successfully applied solution to batch {}", batch);
            true
        } else {
            info!("Not submitting trivial solution for batch {}", batch);
            false
        };
```

**Comment by josojo** [comment]
This PR is no longer necessary, as https://github.com/gnosis/dex-services/pull/296 will just prevent submitting solutions with less objective value

**Pull request closed without merging**